src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationOptions.cs

```diff
@@ -154,24 +156,63 @@ public AzureAppConfigurationOptions UseFeatureFlags(Action<FeatureFlagOptions> c
                     string.Format(ErrorMessages.CacheExpirationTimeTooShort, MinimumFeatureFlagsCacheExpirationInterval.TotalMilliseconds));
             }
 
-            if (!(_kvSelectors.Any(selector => selector.KeyFilter.StartsWith(FeatureManagementConstants.FeatureFlagMarker) && selector.LabelFilter.Equals(options.Label))))
+            if (options.FeatureFlagSelectors.Count() != 0 && options.Label != null)
             {
-                Select(FeatureManagementConstants.FeatureFlagMarker + "*", options.Label);
+                throw new InvalidOperationException($"Please select feature flags by either the {nameof(options.Select)} method or by setting the {nameof(options.Label)}, not both.");
+            }
+            
+            if (options.FeatureFlagSelectors.Count() == 0 && options.Label == null)
+            {
+                // No Select clause and the Label property is not set
+                options.FeatureFlagSelectors.Add(new KeyValueSelector
+                {
+                    KeyFilter = FeatureManagementConstants.FeatureFlagMarker + "*",
+                    LabelFilter = LabelFilter.Null
+                });  
+            }
+            else if (options.FeatureFlagSelectors.Count() == 0 && options.Label != null)
+            {
+                // No Select clause but the Label property is set
+                options.FeatureFlagSelectors.Add(new KeyValueSelector
+                {
+                    KeyFilter = FeatureManagementConstants.FeatureFlagMarker + "*",
+                    LabelFilter = options.Label
+                });
             }
 
+            _featureFlagPrefixes.AddRange(options.FeatureFlagPrefixes);
+
             if (!_adapters.Any(a => a is FeatureManagementKeyValueAdapter))
             {
-                _adapters.Add(new FeatureManagementKeyValueAdapter());
+                _adapters.Add(new FeatureManagementKeyValueAdapter(_featureFlagPrefixes));
             }
 
-            if (!_multiKeyWatchers.Any(kw => kw.Key.Equals(FeatureManagementConstants.FeatureFlagMarker)))
+            foreach (var featureFlagSelector in options.FeatureFlagSelectors)
             {
-                _multiKeyWatchers.Add(new KeyValueWatcher
+                var featureFlagFilter = featureFlagSelector.KeyFilter;
+                var labelFilter = featureFlagSelector.LabelFilter;
+
+                if (!_kvSelectors.Any(selector => selector.KeyFilter == featureFlagFilter && selector.LabelFilter == labelFilter))
                 {
-                    Key = FeatureManagementConstants.FeatureFlagMarker,
-                    Label = options.Label,
-                    CacheExpirationInterval = options.CacheExpirationInterval
-                });
+                    Select(featureFlagFilter, labelFilter);
+                }
+
+                var multiKeyWatcher = _multiKeyWatchers.FirstOrDefault(kw => kw.Key.Equals(featureFlagFilter) && kw.Label.NormalizeNull() == labelFilter.NormalizeNull());
+
```

Comment:
We can optimize so that if the flags are already watched we don't add an additional watcher.

For example,

if .appconfig/featureflag/* is watched, we won't add .appconfig/featureflag/abc*

The interesting part is the cache expiration interval could be different.

Reply:
As discussed offline, this optimization is not worth the effort right now. It's a rare situation that should be optimized on customer side by not registering both `.appconfig/featureflag/*` and `.appconfig/featureflag/abc*` filters.